packages/navigator/src/browser/navigator-tree.ts

```diff
@@ -30,7 +32,102 @@ export class FileNavigatorTree extends FileTree {
     }
 
     async resolveChildren(parent: CompositeTreeNode): Promise<TreeNode[]> {
+        if (FileStatNode.is(parent)) {
+            let resolvedChildren: TreeNode[] = [];
+            let fileStat;
+            if (this.isVirtualRoot(parent)) {
+                fileStat = await this.resolveVirtualRootFileStat(parent);
+                if (fileStat) {
+                    parent.fileStat = fileStat;
+                }
+            } else {
+                fileStat = await this.resolveFileStat(parent);
+            }
+
+            if (fileStat) {
+                resolvedChildren = await this.toNodes(fileStat, parent);
+            }
+            return this.filter.filter(resolvedChildren);
+        }
         return this.filter.filter(super.resolveChildren(parent));
     }
 
+    protected async toNode(fileStat: FileStat, parent: CompositeTreeNode): Promise<FileNode | DirNode> {
+        const id = this.generateFileTreeNodeId(fileStat.uri, parent);
+        const node = this.getNode(id);
+        if (fileStat.isDirectory) {
+            if (DirNode.is(node)) {
+                node.fileStat = fileStat;
+                return node;
+            }
+            return Object.assign(await super.toNode(fileStat, parent), { id });
+        }
+        if (FileNode.is(node)) {
+            node.fileStat = fileStat;
+            return node;
+        }
+        return Object.assign(await super.toNode(fileStat, parent), { id });
+    }
+
+    protected isVirtualRoot(node: CompositeTreeNode): boolean {
+        return !node.visible && DirNode.is(node) && !node.parent;
+    }
+
+    protected generateFileTreeNodeId(uri: string, parent: CompositeTreeNode | undefined): string {
+        if (parent) {
+            const grandparent = parent.parent;
+            if (grandparent && !this.isVirtualRoot(grandparent)) {
+                return this.generateFileTreeNodeId(uri, grandparent);
+            } else {
+                if (FileStatNode.is(parent)) {
+                    if (this.isVirtualRoot(parent)) {
+                        return `${new URI(uri).path.toString()}///${uri}`;
```

Comment:
get rid of code duplication: there should be one single function which accepts a root URI and a node URI and returns an ID. It also should be used from `navigator-model` instead of be reimplemented there.